reark/src/main/java/io/reark/reark/data/store/ContentProviderStore.java

```diff
@@ -76,44 +75,50 @@ protected ContentProviderStore(@NonNull ContentResolver contentResolver) {
 
         updateSubject
                 .onBackpressureBuffer()
-                .observeOn(Schedulers.io())
-                .subscribe(pair -> {
-                    updateIfValueChanged(this, pair);
-                });
+                .doOnNext(pair -> locker.createLock(pair.second))
+                .flatMap(pair -> updateIfValueChanged(this, pair.first, pair.second))
+                .subscribe(locker::freeLock, error -> Log.e(TAG, "Error updating", error));
     }
 
-    private static <T> void updateIfValueChanged(ContentProviderStore<T> store, Pair<T, Uri> pair) {
-        final Cursor cursor = store.contentResolver.query(pair.second, store.getProjection(), null, null, null);
-        T newItem = pair.first;
-        boolean valuesEqual = false;
-
-        if (cursor != null) {
-            if (cursor.moveToFirst()) {
-                T currentItem = store.read(cursor);
-                valuesEqual = newItem.equals(currentItem);
-
-                if (!valuesEqual) {
-                    Log.v(TAG, "Merging values at " + pair.second);
-                    newItem = store.mergeValues(currentItem, newItem);
-                    valuesEqual = newItem.equals(currentItem);
-                }
-            }
-            cursor.close();
-        }
-
-        if (valuesEqual) {
-            Log.v(TAG, "Data already up to date at " + pair.second);
-            return;
-        }
-
-        final ContentValues contentValues = store.getContentValuesForItem(newItem);
-
-        if (store.contentResolver.update(pair.second, contentValues, null, null) == 0) {
-            final Uri resultUri = store.contentResolver.insert(pair.second, contentValues);
-            Log.v(TAG, "Inserted at " + resultUri);
-        } else {
-            Log.v(TAG, "Updated at " + pair.second);
-        }
+    private static <T> Observable<Uri> updateIfValueChanged(ContentProviderStore<T> store, T item, Uri uri) {
+        return Observable.just(item)
+                .observeOn(Schedulers.io())
+                .map(newItem -> {
+                    synchronized (store.locker.getLock(uri)) {
```

Comment:
#134 continues from this, decided to implement this there with semaphores instead of a large synchronized block.

Integration tests for stores would be nice..